code/obj/mining_cloud_storage.dm

```diff
@@ -0,0 +1,372 @@
+				continue
+			if(M.material.name != M.initial_material_name)
+				continue
+			if (O.loc == user)
+				continue
+			if (O in user.contents)
+				continue
+			src.load_item(M)
+			M.set_loc(src)
+			playsound(get_turf(src), sound_load, 40, 1)
+			sleep(0.5)
+			if (user.loc != staystill) break
+		boutput(user, "<span class='notice'>You finish stuffing [O] into [src]!</span>")
+		return
+
+	attackby(obj/item/W as obj, mob/user as mob)
+		if (istype(W, /obj/item/raw_material/) && src.accept_loading(user))
+			var/obj/item/raw_material/R = W
+			if(R.material.name != R.initial_material_name)
+				boutput(user, "<span class='alert'>[src] rejects the anomalous ore.</span>")
+				return
+			user.visible_message("<span class='notice'>[user] loads [W] into the [src].</span>", "<span class='notice'>You load [W] into the [src].</span>")
+			src.load_item(W,user)
+
+		else
+			src.health = max(src.health-W.force,0)
+			src.check_health()
+			..()
+
+	proc/check_health()
+		if(!src.health && !broken)
+			src.broken = 1
+			src.visible_message("<span class='alert'>[src] breaks!</span>")
+			src.icon_state = "ore_storage_unit-broken"
+
+	proc/load_item(var/obj/item/raw_material/R,var/mob/living/user)
+		if (!R)
+			return
+		var/amount_loaded = 0
+		if(R.amount == 1)
+			R.set_loc(src)
+			amount_loaded++
+			if (user && R)
+				user.u_equip(R)
+				R.dropped()
+		else if(R.amount>1)
+			R.set_loc(src)
+			for(R.amount,R.amount > 0, R.amount--)
+				var/obj/item/raw_material/new_mat = unpool(R.type)
+				new_mat.loc = src
+				amount_loaded++
+			if (user && R)
+				user.u_equip(R)
+				R.dropped()
+			pool(R)
+		update_ore_amount(R.material_name,amount_loaded)
+
+
+	proc/accept_loading(var/mob/user,var/allow_silicon = 0)
+		if (!user)
+			return 0
+		if (src.status & BROKEN || src.status & NOPOWER)
+			return 0
+		if (!istype(user, /mob/living/))
+			return 0
+		if (istype(user, /mob/living/silicon) && !allow_silicon)
+			return 0
+		var/mob/living/L = user
+		if (!isalive(L) || L.transforming)
+			return 0
+		return 1
+
+	proc/update_ore_amount(var/material_name,var/delta)
+		if(ores[material_name])
+			var/datum/ore_cloud_data/OCD = ores[material_name]
+			OCD.amount += delta
+			OCD.amount = max(OCD.amount,0)
+			ores[material_name] = OCD
+		else if (delta > 0)
+			ores += material_name
+			var/datum/ore_cloud_data/OCD = new /datum/ore_cloud_data
+			OCD.amount += delta
+			OCD.for_sale = 0
+			OCD.price = 0
+			ores[material_name] = OCD
+
+	proc/update_ore_for_sale(var/material_name,var/new_for_sale)
+		if(ores[material_name])
+			var/datum/ore_cloud_data/OCD = ores[material_name]
+			if(isnull(new_for_sale))
+				OCD.for_sale = !OCD.for_sale
+			else
+				OCD.for_sale = new_for_sale
+			ores[material_name] = OCD
+		return
+
+	proc/update_ore_price(var/material_name,var/new_price)
+		if(ores[material_name])
+			var/datum/ore_cloud_data/OCD = ores[material_name]
+			OCD.price = max(0,new_price)
+			ores[material_name] = OCD
```

Comment:
```suggestion
```
See comment about datums being passed by reference.

Reply:
resolved in pending commit